Add possibility for miniz to use an external crc definition.

jjlilley opened the pull request:
Summary:
We add an #ifdef check for USE_EXTERNAL_MZCRC, in which case miniz
will look for an external mz_crc32 definition. The default behavior
is unchanged.

Test Plan: Unchanged default behavior, but buck test caffe2/test/...

Differential Revision: D17814440

resistor commented:
Can we propose this upstream?

resistor commented:
Upstream being https://github.com/richgel999/miniz

jjlilley commented:
> Upstream being https://github.com/richgel999/miniz

Ah good idea. I didn't realize that we kept these tightly synced. Just sent a pull request to that repo:
   https://github.com/richgel999/miniz/pull/141/

resistor commented:
> Ah good idea. I didn't realize that we kept these tightly synced. Just sent a pull request to that repo:
> [richgel999/miniz#141](https://github.com/richgel999/miniz/pull/141)

We don't keep them tightly synced, but I'd like to protect against this change getting accidentally overwritten if we re-sync in the future.

resistor commented on third_party/miniz-2.0.8/miniz.c:
```diff
@@ -82,6 +82,20 @@ mz_ulong mz_adler32(mz_ulong adler, const unsigned char *ptr, size_t buf_len)
         }
         return ~crcu32;
     }
+#elif defined(USE_EXTERNAL_MZCRC)
```
Should this be `#elif defined(USE_EXTERNAL_MZCRC) && USE_EXTERNAL_MZCRC`?

jjlilley replied:
Inside this specific file (as well as the .h, which widely uses ifdefs), defined(X) seems more idiomatic. This would be the only case of 'defined(X) && X' pattern, so it seems best to follow the rest of the file.

jjlilley commented:
fwiw, I just updated this revision to exactly match the pull request that I sent to richgel999/miniz#141

jjlilley commented:
The identical upstream pull request in miniz was just accepted and landed:
   https://github.com/richgel999/miniz/pull/141

Could somebody accept this?

Commit:
Add possibility for miniz to use an external crc definition. (#27558)

Summary:
We add an #ifdef check for USE_EXTERNAL_MZCRC, in which case miniz
will look for an external mz_crc32 definition. The default behavior
is unchanged.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/27558

Test Plan: Unchanged default behavior, but buck test caffe2/test/...

Differential Revision: D17814440

Pulled By: jjlilley

fbshipit-source-id: 002f27e4f282bc09376fa6d614b4802e555ea782
